🤖 Create the HCB Discord bot

YodaLightsabr opened the pull request:
## Summary of the problem

There aren't many ways for users to engage with HCB outside of https://hcb.hackclub.com. 


## Describe your changes

This PR makes it easy for our users, particularly teens and robotics teams, to manage their funds in a new way.

polypixeldev committed:
Add webhook controller and verify key

polypixeldev committed:
Add linking HCB to Discord user

YodaLightsabr committed:
Merge branch 'main' into discord-bot

YodaLightsabr committed:
Add command handler and command registration

YodaLightsabr committed:
Fix routes.rb

YodaLightsabr committed:
Add command routing in the Discord controller

YodaLightsabr committed:
Merge branch 'main' into discord-bot

YodaLightsabr committed:
Abstract command handling logic into ::Discord::HandleInteractionJob

YodaLightsabr committed:
Discord

YodaLightsabr committed:
Add server linking and public activity to Discord bot

Co-authored-by: rluo.dev <rluodev@users.noreply.github.com>

YodaLightsabr committed:
Transactions

YodaLightsabr committed:
Lint

YodaLightsabr committed:
Merge branch 'main' into discord-bot

garyhtou commented on app/jobs/discord/handle_interaction_job.rb:
```diff
@@ -0,0 +1,231 @@
+      elsif !@current_event.present? && @current_user.present?
+        respond content: "You've linked your Discord and HCB accounts, but this Discord server isn't connected to an HCB organization yet:", components: button_to("Set up HCB on this server", url_helpers.discord_setup_url(guild_id: @guild_id, channel_id: @channel_id)), embeds: linking_embed
+      elsif @current_event.present? && !@current_user.present?
+        respond content: "HCB has already been setup for this Discord server, but your Discord account isn't linked to your HCB account yet:", components: button_to("Link Discord account", url_helpers.discord_link_url(discord_id: @user_id)), embeds: linking_embed
```
```suggestion
        respond content: "This Discord server is connected to #{@current_event.name} on HCB. HCB is the platform your team uses to manage its finances. Finish your setup by linking your Discord account to HCB:", components: button_to("Link Discord account", url_helpers.discord_link_url(discord_id: @user_id)), embeds: linking_embed
```

garyhtou commented on app/jobs/discord/handle_interaction_job.rb:
```diff
@@ -0,0 +1,231 @@
+      elsif @current_event.present? && !@current_user.present?
+        respond content: "HCB has already been setup for this Discord server, but your Discord account isn't linked to your HCB account yet:", components: button_to("Link Discord account", url_helpers.discord_link_url(discord_id: @user_id)), embeds: linking_embed
+      else
+        respond content: "Link your HCB account, and then connect this Discord server to an HCB organization:", components: [button_to("Link Discord account", url_helpers.discord_link_url(discord_id: @user_id)), button_to("Set up HCB on this server", url_helpers.discord_setup_url(guild_id: @guild_id, channel_id: @channel_id))], embeds: linking_embed
```
URLs should be signed.

Also, should we add a `1.` and `2.` to the buttons so people do it in the order we prefer?

garyhtou commented on app/jobs/discord/handle_interaction_job.rb:
```diff
@@ -0,0 +1,231 @@
+      end
+    end
+
+    def setup_command
+      link_command
+    end
```
Is this used?

garyhtou commented on app/jobs/discord/handle_interaction_job.rb:
```diff
@@ -0,0 +1,231 @@
+    def transactions_command
+      return require_linked_event unless @current_event
+
+      transactions = @current_event.canonical_pending_transactions.order(created_at: :desc).limit(4)
```
We should probably use this to get the link of transactions:

https://github.com/hackclub/hcb/blob/ca3cecee49ca389b1d652e7613c204804bfaecc2/app/controllers/events_controller.rb#L189-L203

This is because CPT is incomplete and will miss some transactions

garyhtou replied:
done

garyhtou commented on app/jobs/discord/handle_interaction_job.rb:
```diff
@@ -0,0 +1,231 @@
+
+      transactions = @current_event.canonical_pending_transactions.order(created_at: :desc).limit(4)
+
+      if transactions.length == 0
```
```suggestion
      if transactions.none?
```

garyhtou commented on app/jobs/discord/process_notification_job.rb:
```diff
@@ -0,0 +1,57 @@
+
+    private
+
+    def bot
+      @bot ||= Discordrb::Bot.new token: Credentials.fetch(:DISCORD__BOT_TOKEN)
+    end
+
+    def color
+      if Rails.env.development?
+        0x33d6a6
+      else
+        0xec3750
+      end
+    end
```
I'd recommend extracting this out to `Discord::Bot`. It would handle it:
- Discordrb client
- Color
- and Faraday client

garyhtou commented on app/models/ahoy/visit.rb:
```diff
@@ -33,8 +33,7 @@
 #
-#  index_ahoy_visits_on_user_id      (user_id)
-#  index_ahoy_visits_on_visit_token  (visit_token) UNIQUE
+#  index_ahoy_visits_on_user_id  (user_id)
```
Discard changes.
```suggestion
#  index_ahoy_visits_on_user_id      (user_id)
#  index_ahoy_visits_on_visit_token  (visit_token) UNIQUE
```

garyhtou commented on app/models/event.rb:
```diff
@@ -38,13 +38,17 @@
 #
 # Indexes
 #
+#  index_events_on_discord_channel_id   (discord_channel_id) UNIQUE
+#  index_events_on_discord_guild_id     (discord_guild_id) UNIQUE
```
It's good that we have this unique constraint in the DB. But let's also add Rails validations for them

garyhtou replied:
Done

garyhtou commented on app/models/event.rb:
```diff
@@ -888,6 +892,10 @@ def point_of_contact_history
                                   .filter_map { |(old_id, _new_id)| User.find_by(id: old_id) }
   end
 
+  def linked_discord_guild?
+    discord_guild_id.present?
+  end
```
We rename this and it's usages to `has_discord_guild?`?
```suggestion
  def has_discord_guild?
    discord_guild_id.present?
  end
```

garyhtou replied:
[`cf97b36` (#11791)](https://github.com/hackclub/hcb/pull/11791/commits/cf97b36c41887cc6ba22284da3bbfd642bbe177d)

garyhtou commented on app/models/user.rb:
```diff
@@ -474,6 +476,18 @@ def managed_active_teenagers_count
     User.active_teenager.joins(organizer_positions: :event).where(events: { id: managed_events }).distinct.count
   end
 
+  def linked_discord_account?
+    discord_id.present?
+  end
```
Rename it's usages too
```suggestion
  def has_discord_account?
    discord_id.present?
  end
```

garyhtou replied:
[`cf97b36` (#11791)](https://github.com/hackclub/hcb/pull/11791/commits/cf97b36c41887cc6ba22284da3bbfd642bbe177d)

garyhtou commented on app/views/discord/link.html.erb:
```diff
@@ -0,0 +1,37 @@
+  <div class="flex items-start">
+    <span class="muted mr2"><%= inline_icon "important" %></span>
+
+    <p class="mt0">This Discord user will be able to see information about your organizations, but it cannot take any action on its own.</p>
```
```suggestion
    <p class="mt0">This Discord user will be able to see information about your organizations, but it cannot perform any action on its own.</p>
```

garyhtou commented on app/views/discord/setup.html.erb:
```diff
@@ -0,0 +1,48 @@
+  <div class="flex items-start">
+    <span class="muted mr2"><%= inline_icon "important" %></span>
+
+    <p class="mt0">HCB will send notifications about your organization to the channel you ran the command in.</p>
```
```suggestion
    <p class="mt0">
      HCB will send notifications about your organization to the <strong><%= @raw_ch_response["name"] %>.
      <br/>
      <small>Want to choose a different channel? Re-run the <code>/link</code> command in your desired channel.</small>
    </p>
```

garyhtou commented on app/views/discord/setup.html.erb:
```diff
@@ -0,0 +1,48 @@
+<div class="flex items-center justify-center g1">
+  <%= form_with url: discord_create_server_link_path, method: :post, class: "flex items-center justify-center g1" do |f| %>
+    <select name="event_id" class="form-select">
+      <% current_user.organizer_positions.each do |position| %>
+        <option value="<%= position.event.id %>"><%= position.event.name %></option>
+      <% end %>
```
```suggestion
      <% current_user.events.each do |event| %>
        <option value="<%= event.id %>"><%= event.name %></option>
      <% end %>
```

garyhtou reviewed (commented):
Comments from synchronous review with @polypixeldev and @YodaLightsabr

garyhtou committed:
Directly access event

garyhtou committed:
Copy edit for link server flow

garyhtou committed:
Copy edit for user link flow

garyhtou committed:
Add comment to explain magic number/flags

YodaLightsabr committed:
Use # for channel

YodaLightsabr committed:
Allow multiple or 1 components to be passed in and refactor button_to to be one component

garyhtou committed:
Add guard clause to handle discord pings

garyhtou committed:
Add comment to explain magic number/flag

YodaLightsabr committed:
Create DiscordPolicy

YodaLightsabr committed:
Remove `put` calls

YodaLightsabr committed:
`@current_user` → `@user`

garyhtou committed:
Remove unnecessary `require`s

garyhtou committed:
[DiscordController] Rename actions to be suffixed with `_webhook`

- `webhook` -> `event_webhook`
- `interaction` -> `interaction_webhook`

garyhtou commented on app/controllers/discord_controller.rb:
```diff
@@ -0,0 +1,169 @@
+  before_action :verify_discord_signature, only: [:webhook, :interaction]
+  skip_after_action :verify_authorized
+
+  def webhook
```
[`0eb4553` (#11791)](https://github.com/hackclub/hcb/pull/11791/commits/0eb4553e43fcd5e85090a30e33c0a58e8fbb32db)

garyhtou committed:
Add comment to explain magic number/flag

garyhtou committed:
Rename `linked_discord_*?` to `has_discord_*?`

- `linked_discord_guild?` -> `has_discord_guild?`
- `linked_discord_account?` -> `has_discord_account?`

garyhtou committed:
Add uniqueness validations for Discord ids

garyhtou committed:
[Uniqueness validation] Add nicer validation messages

garyhtou committed:
`verify_authorized` after `DiscordController#link`

rluodev committed:
verify signatures for links

garyhtou committed:
[DiscordController] Add improved error handling

garyhtou committed:
Fix typo (`validate` -> `validates`)

garyhtou commented on app/controllers/discord_controller.rb:
```diff
@@ -0,0 +1,169 @@
+
+  def create_server_link
+    event = Event.find(params[:event_id])
+    unless current_user.organizer_positions.exists?(event:)
```
Handled by Ian

garyhtou commented on app/controllers/discord_controller.rb:
```diff
@@ -0,0 +1,169 @@
+      redirect_to root_path and return
+    end
+
+    event.update!(discord_guild_id: params[:guild_id], discord_channel_id: params[:channel_id])
```
done

garyhtou commented on app/controllers/discord_controller.rb:
```diff
@@ -0,0 +1,169 @@
+    unless current_user.organizer_positions.exists?(event:)
+      flash[:error] = "We could not unlink your organization from your Discord server"
+      redirect_to root_path and return
+    end
```
Handled by ian

garyhtou committed:
Copy edit for unlinking server

garyhtou commented on app/controllers/discord_controller.rb:
```diff
@@ -0,0 +1,169 @@
+
+    event.update!(discord_guild_id: nil, discord_channel_id: nil)
+
+    bot.send_message(cid, "The HCB organization #{event.name} has been unlinked from this Discord server, and notifications/announcements will no longer be sent here.")
```
done 40a5407d8391b159ed25dcf4177610bdf2d16fda

garyhtou committed:
Make it more clear that you're in development

garyhtou commented on app/jobs/discord/handle_interaction_job.rb:
```diff
@@ -0,0 +1,231 @@
+          backtrace = "#{backtrace[0..4_000]}..."
+        end
+
+        respond content: "**That didn't work!**\nWe're going to debug what went wrong.", embeds: [
```
done 7bcfe9e2f6524e100125e2c50a00f640d7957edc

garyhtou committed:
Copy edit for user link instruction in discord message

garyhtou committed:
Remove unused method

garyhtou commented on app/jobs/discord/process_notification_job.rb:
```diff
@@ -0,0 +1,57 @@
+
+      text = ReverseMarkdown.convert(html)[0..4000]
+
+      bot.send_message(@event.discord_channel_id, nil, false, { description: text, timestamp: Time.at((timestamp.to_i / 1000).to_i).iso8601, author: { name: user_name, icon_url: user_avatar_url }, color: })
```
Assigned to Ian

alexdevz committed:
Make most link/unlink and receipts ephemeral

YodaLightsabr committed:
Refactor process notification job

polypixeldev committed:
Redirect users to install link if discord ID is nil or invalid

polypixeldev committed:
Move install link redirect to if signed message is nil

polypixeldev committed:
Change setup to use Discordrb to fetch

polypixeldev committed:
Pass allow_other_host: true to redirects

YodaLightsabr committed:
Support ephemeral commands

YodaLightsabr committed:
Add back setup command

YodaLightsabr committed:
Update app/jobs/discord/handle_interaction_job.rb

Co-authored-by: Gary Tou <gary@garytou.com>

YodaLightsabr committed:
Use Discordrb constants

garyhtou committed:
Rewrite Discord ID signing code

YodaLightsabr committed:
Update app/jobs/discord/handle_interaction_job.rb

Co-authored-by: Gary Tou <gary@garytou.com>

YodaLightsabr committed:
Fix DiscordPolicy

garyhtou committed:
Remove unnecessary return

YodaLightsabr committed:
`@current_user` → `@user`

garyhtou committed:
Fix typo

garyhtou committed:
Fix validation to allow nil

YodaLightsabr committed:
Always permit the setup route

garyhtou committed:
Fix typo again

YodaLightsabr committed:
Copy edit

YodaLightsabr committed:
Redirect back or to

YodaLightsabr committed:
Bot module

garyhtou committed:
Refactor redirects to discord install link

Makes the code slightly more readable and also add `and return` to ensure we stop processing the request

garyhtou commented on app/controllers/discord_controller.rb:
```diff
@@ -0,0 +1,197 @@
+  def link
+    authorize nil, policy_class: DiscordPolicy
+    @signed_discord_id = params[:signed_discord_id]
+    redirect_to(install_link, allow_other_host: true) if @signed_discord_id.nil?
```
[`7d02c1b` (#11791)](https://github.com/hackclub/hcb/pull/11791/commits/7d02c1bdbb398269cf4a4468c2d055322598c6df)

garyhtou committed:
Only redirect to event if it's present

garyhtou committed:
Remove `puts` messages

garyhtou committed:
Remove unused Faraday Discord API call

garyhtou committed:
Sub Faraday API call for Discordrb client

garyhtou committed:
Make `InvalidSignature` error more generic

YodaLightsabr committed:
Use Discord::Bot

garyhtou committed:
[Transactions command] Use `TransactionGroupingEngine` instead of just CPT

garyhtou committed:
Use `@current_event` instead of `@event`

garyhtou committed:
Discard changes to Ahoy::Visit

garyhtou commented on app/controllers/discord_controller.rb:
```diff
@@ -0,0 +1,169 @@
+      redirect_to root_path and return
+    end
+
+    event.update!(discord_guild_id: params[:guild_id], discord_channel_id: params[:channel_id])
```
Assigned to @rluodev

garyhtou replied:
done

garyhtou committed:
Add TODO

rluodev committed:
fix error handling

garyhtou commented on app/controllers/discord_controller.rb:
```diff
@@ -0,0 +1,169 @@
+  end
+
+  def create_link
+    current_user.update!(discord_id: params[:discord_id])
```
done

YodaLightsabr committed:
Interactions

garyhtou committed:
Verify can access Discord user when linking it

garyhtou committed:
Extract methods to `Discord::Bot`

garyhtou committed:
Remove unrelated comment

garyhtou committed:
lint

garyhtou committed:
Pull Discord server icon URL directly from Discordrb

garyhtou committed:
Reuse Faraday connection

garyhtou committed:
Bring back logic to handle servers without custom icons

garyhtou committed:
Patch "channel is within guild" check

garyhtou committed:
Merge branch 'main' into discord-bot

garyhtou committed:
lint

YodaLightsabr committed:
Handle interactions properly and make linking embed ephemeral always

garyhtou committed:
[`/transactions`] Indicate pending transactions

YodaLightsabr committed:
Add create reimbursement button and correct components array _before_ returning passthrough response

YodaLightsabr committed:
Sort reimbursement reports

YodaLightsabr committed:
Add blankslate

YodaLightsabr committed:
Remove problematic code
